### ISSUE TITLE
issues with Unicode in query string

### ISSUE THREAD

**Issue by ninegua** [cheecheeo/haskell-cgi#23]
~~~
*Main> import Network.CGI
*Main Network.CGI> import Data.ByteString.Char8 as BS
*Main Network.CGI BS> let [(n,v)] = formDecode "term=%CE%BBx.x"
*Main Network.CGI BS> BS.pack v
"\187x.x"
*Main Network.CGI BS> v
"\955x.x"
*Main Network.CGI BS> putStrLn v
λx.x
~~~

The above shows what is wrong with `BS.pack` [in this line](https://github.com/cheecheeo/haskell-cgi/blob/master/Network/CGI/Protocol.hs#L172): it is not respecting Unicode strings that are decoded from `QUERY_STRING` environment. The consequence is that `getInput` is no longer returning the correct value, and it is impossible to recover the correct value afterwards.

Older bytestring package (e.g., bytestring-0.9.2.1) actually was able to pack Unicode into UTF-8 sequences even when you use Char8, but this is no longer the case. I would advise the CGI package either store things internally as String, or use utf8-string package to pack things properly.

**Comment by ninegua** [comment]
I'm actually wrong about the older versions of bytestring claim. So please ignore that. I was actually trying older versions of cgi and network packages, where the `formDecode` won't treat input as UTF-8. So the decoded string was strictly 8-bit for each character, and therefore would pack/unpack without loss.

**Comment by ninegua** [comment]
Any help to have the current issue resolved would be much appreciated! I can provide a merge request if necessary. Please let me know which way is preferred, whether changing `inputValue` of `Input` data type to String, or use conversion functions from `utf8-string` instead of `BS.unpack`. Thanks!

**Comment by cheecheeo** [comment]
@ninegua For now I'd prefer the `String` route. If you have time, a doctest-style unit test would be awesome to add, don't worry about hooking it in to cabal or a test framework unless you really want to.

**Comment by ninegua** [comment]
Finally got some time to work on this. It turns out that changing `inputValue` to String type will affect some other parts that expect inputValue to be ByteString, specifically, the handling of post data. So I opted the second approach, i.e., using `fromString` from utf8-string package instead of `BS.pack`. 

My pull request is https://github.com/cheecheeo/haskell-cgi/pull/24. I attempted to include a doctest style comment, but when I tried to use doctest, it complained about some CPP stuff despite that I gave it -XCPP. So I wasn't able to successfully test it. That being said, manually typing the example into GHCi gives the right output for me (It was wrong output before this patch).

**Comment by ninegua** [comment]
BTW, another solution is to use a ByteString equivalent of getEnvironment, which is provided in recent unix package for Posix systems `System:

https://hackage.haskell.org/package/unix-2.7.1.0/docs/System-Posix-Env-ByteString.html

Unfortunately this package doesn't support Windows.

**Comment by cheecheeo** [comment]
@ninegua Thanks for the PR. Based on your long description in the commit the problem your dealing with is due to the `unEscapeString` function. The behavior of `unEscapeString` has changed on the 'next' branch, specifically, it no longer does UTF-8 decoding. Can you check out that branch and see if you're still seeing the issue with that code? If you are, can you base your change off of that branch?

Thank you.

**Comment by ninegua** [comment]
Thanks @cheecheeo . I took a look at the `next` branch, and saw this commit https://github.com/cheecheeo/haskell-cgi/commit/42f4ca7dab1afffeb76beeff13f6158124c59181 actually fixed the same issue. So when are we going to see this branch getting merged and released?

**Comment by cheecheeo** [comment]
@ninegua I'll cut a release and upload it to hackage either this weekend or next. If you're using `stack` you can reference that commit directly in your `stack.yaml` file as documented here: http://docs.haskellstack.org/en/stable/yaml_configuration/#project-config

I'll close this ticket when I've completed the release.

**Comment by cheecheeo** [comment]
Merged 7f4ad94aaa97468b7e39925b8040781074ba7414 hackage: http://hackage.haskell.org/package/cgi-3001.3.0.0

**Comment by ninegua** [comment]
Glad to see this gets merged to master and released. Thanks!
